### PR TITLE
Tests/coverage plugin integrations tests for review

### PR DESCRIPTION
**Pull request by Tjuri**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by uhafner** [comment]
Here is a patch for your delta computation:
https://github.com/jenkinsci/code-coverage-api-plugin/commit/4f881723efad6fd313b25b042088aa6e520d74f0

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+    @Test
+    public void pipelineFailWhenCoverageDecreases() {
+
+        // TODO: fails
+        Run<?, ?> firstBuild = createPipelineJobAndAssertBuildResult(
+                "node {"
+                        + "   publishCoverage adapters: [jacocoAdapter('**/*.xml')], failBuildIfCoverageDecreasedInChangeRequest: true"
+                        + "}", Result.SUCCESS, CoveragePluginITestUtil.JACOCO_CODING_STYLE_FILE_NAME);
+        Run<?, ?> secondBuild = createPipelineJobAndAssertBuildResult(
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java:181]: This creates another job. However it must be the second build of the same job!

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+        // TODO: doesnt work yet
+        Run<?, ?> firstBuild = createPipelineJobAndAssertBuildResult(
+                "node {"
+                        + "   publishCoverage adapters: [jacocoAdapter(path: '*.xml')], sourceFileResolver: sourceFiles('NEVER_STORE')"
+                        + "}", Result.SUCCESS, CoveragePluginITestUtil.JACOCO_CODING_STYLE_FILE_NAME, CoveragePluginITestUtil.JACOCO_CODING_STYLE_DECREASED_FILE_NAME);
+
+        Run<?, ?> secondBuild = createPipelineJobAndAssertBuildResult(
+                "node {"
+                        + "   publishCoverage adapters: [jacocoAdapter(path: '*.xml')], sourceFileResolver: sourceFiles('NEVER_STORE')"
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java:279]: Here you need to call

```

```

**Review by uhafner — COMMENTED** [review]
It is hard to make a review here when almost every line contains a warning. Please check the coding style.

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+    public void freestyleZeroReportsFail() {
+        FreeStyleProject project = createFreeStyleProject();
+
+        CoveragePublisher coveragePublisher = createPublisherWithJacocoAdapter("*.xml");
+        coveragePublisher.setFailNoReports(true);
+        project.getPublishersList().add(coveragePublisher);
+
+        Run<?, ?> build = buildWithResult(project, Result.FAILURE);
+
+        assertThat(build.getResult()).isEqualTo(Result.FAILURE);
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java:209]: This is already included in line 207

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+
+    @Test
+    public void freestyleZeroReportsOkay() {
+        FreeStyleProject project = createFreeStyleProject();
+
+        CoveragePublisher coveragePublisher = createPublisherWithCoberturaAdapter("*.xml");
+        coveragePublisher.setFailNoReports(false);
+        project.getPublishersList().add(coveragePublisher);
+
+        Run<?, ?> build = buildWithResult(project, Result.SUCCESS);
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java:220]: DRY!

Try to use one method with parameters.

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        copyFilesToWorkspace(project, CoveragePluginITestUtil.COBERTURA_COVERAGE_FILE_NAME);
+
+        CoveragePublisher coveragePublisher = createPublisherWithCoberturaAdapter("*.xml");
+        coveragePublisher.setSkipPublishingChecks(false);
+
+        project.getPublishersList().add(coveragePublisher);
+        Run<?, ?> build = buildSuccessfully(project);
+
+        assertThat(build.getLog(1000))
+                .contains("[Checks API] No suitable checks publisher found.");
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginFreestyleITest.java:348]: See above: almost the same code

File: src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        copySingleFileToAgentWorkspace(agent, job, CoveragePluginITestUtil.JACOCO_ANALYSIS_MODEL_FILE_NAME,
+                CoveragePluginITestUtil.JACOCO_ANALYSIS_MODEL_FILE_NAME);
+
+        Run<?, ?> build = buildSuccessfully(job);
+
+        CoveragePluginITestUtil.assertLineCoverageResultsOfBuild(
+                Collections.singletonList(CoveragePluginITestUtil.JACOCO_ANALYSIS_MODEL_LINES_TOTAL),
+                Collections.singletonList(CoveragePluginITestUtil.JACOCO_ANALYSIS_MODEL_LINES_COVERED),
+                build);
+    }
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginPipelineITest.java:315]: Here assertions are missing that check for the source code.

**Review by uhafner — CHANGES_REQUESTED** [review]

**Comment by NikolasFunction** [comment]
Unfortunately, we couldn't achieve a green tick here. Thus, a few notes about our work here:

- Some integration tests revealed bugs in the source code. These tests are ignored using `@Ignore`
- Integration test `pipelineWithDockerAgent` currently fails when executed on a windows machine during the pipeline. We couldn't find any reason why.
- Static code analysis notices a few issues. We could not avoid them due to the following reason(s):
  - Methods throw Exceptions because inside, a used method throws `Exception` explicitly
  - Unused imports are actually used, but not in the working code itself but in the javadoc. Is this a bug in static code analysis?
  - `JavaGitContainer` is not our responsibility

**Review by uhafner — APPROVED** [review]
Thank you very much for these tests! Since we can merge only of these similar pull requests of our assignment I am closing this one now.

**Pull request closed without merging**